File: nvpl_blas/c/dsymv.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_dsymv
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t N = 2;
    nvpl_int_t lda = 2;
    nvpl_int_t incX = 1;
    nvpl_int_t incY = 2;
    double alpha = 1.0;
    double beta = 1.0f;
    enum CBLAS_ORDER order = CblasRowMajor;
    enum CBLAS_UPLO uplo = CblasUpper;
    double * A = NULL;
    double * X = NULL;
    double * Y = NULL;
    nvpl_int_t len_x = 1 + (N - 1) * labs(incX);
    nvpl_int_t len_y = 1 + (N - 1) * labs(incY);

    printf("\nExample: cblas_dsymv for computing a matrix-vector product for a symmetric matrix.\n\n");
    printf("#### args: n=%" PRId64 ", lda=%" PRId64 ", incX=%" PRId64 ", incY=%" PRId64 ", order=%c, uplo=%c\n",
           (int64_t)N, (int64_t)lda, (int64_t)incX, (int64_t)incY, order_to_char(order), uplo_to_char(uplo));
    printf("           alpha=%g, beta=%g\n", alpha, beta);

    // allocate memory
    A = (double *)malloc(lda * N  * sizeof(double));
    X = (double *)malloc(len_x * sizeof(double));
    Y = (double *)malloc(len_y * sizeof(double));

    // fill data
    fill_dmatrix(A, N, N, lda, order, UPLO_TO_FILL_MODE(uplo), CblasNonUnit);
    fill_dvector(X, N, incX);
    fill_dvector(Y, N, incY);

    // print input data
    print_dmatrix(A, N, N, lda, "A", order);
    print_dvector(X, N, incX, "X");
    print_dvector(Y, N, incY, "Y");

    // call cblas_dsymv
    cblas_dsymv(order, uplo, N, alpha, A, lda, X, incX, beta, Y, incY);

    // print result
    printf("\nAfter dsymv operation:\n");
    print_dvector(Y, N, incY, "Y");

    // release memory
    free(A);
    free(X);
    free(Y);
    return 0;
}

File: nvpl_blas/f77/dsymv.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     DSYMV 
!
!******************************************************************************
      program   DSYMV_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

*       Intrinsic Functions
      intrinsic  abs

      character*1     uplo
      parameter       (uplo='U')
      integer         n, lda, incx, incy, i
      parameter       (n=2, lda=2, incx=1, incy=1)
      double precision alpha, beta
      parameter       (alpha=1.0, beta=1.0)
      integer         xsize, ysize
      parameter       (xsize=1+(n-1)*abs(incx),
     &                 ysize=1+(n-1)*abs(incy))
      double precision x(xsize), y(ysize), a(lda,n)

*       External Subroutines
      external        print_dvector, fill_dvector,
     &                print_dmatrix, fill_dmatrix
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, n, lda, incx, incy, uplo
      print 101, alpha, beta
      call fill_dvector(x,n,incx)
      call fill_dvector(y,n,incy)
      call fill_dmatrix(a,n,n,lda,uplo,'N')
      call print_dmatrix(a,n,n,lda,'A')
      call print_dvector(x,n,incx,'X')
      call print_dvector(y,n,incy,'Y')
*
*      Call DSYMV subroutine
      call DSYMV(uplo,n,alpha,a,lda,x,incx,beta,y,incy)
*
      print*
      print 102
      call print_dvector(y,n,incy,'Y')

      stop
 99   format('Example: DSYMV for computing a matrix-vector product'
     &       ' for a symmetric matrix.')
 100  format('#### args: n=',i1,', lda=',i1,', incx=',i1,', incy=',i1,
     &       ', uplo=',a1)
 101  format(11x,'alpha=',f4.2,' beta=',f4.2)
 102  format('The result of DSYMV: ')
      end
